[DataPipe] Add function for deprecation of functional DataPipe names

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78970

Fixes https://github.com/pytorch/data/issues/480.

For the list of deprecated DataPipes (or functional names), see https://github.com/pytorch/data/issues/163

Testing:
```python
IterableWrapper(range(10)).open_file_by_iopath()
```
Returns:
```
/Users/.../pytorch/torch/utils/data/datapipes/utils/common.py:171: FutureWarning: `IoPathFileOpener()`'s functional API `.open_file_by_iopath()` is deprecated since 1.12 and will be removed in 1.14.
See https://github.com/pytorch/data/issues/163 for details.
Please use `.open_files_by_iopath()` instead.
  warnings.warn(msg, FutureWarning)
  ```

NivekT committed:
[DataPipe] Add function for deprecation of functional DataPipe names

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -104,6 +104,23 @@ def validate_pathname_binary_tuple(data: Tuple[str, IOBase]):
         )
 
 
+# Deprecated function names and its corresponding kwargs for the `_deprecation_warning` function
+_deprecated_functional_names: Dict[str, Dict] = {"open_file_by_fsspec":
```
Going forward, we might need `_deprecated_functional_names` for both `IterDataPipe` and `MapDataPipe`.

NivekT replied:
I'm going to add a string to indicate whether it is for `IterDataPipe` or `MapDataPipe`.

ejguan commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -109,6 +110,9 @@ class IterDataPipe(IterableDataset[T_co], metaclass=_IterDataPipeMeta):
 
     def __getattr__(self, attribute_name):
         if attribute_name in IterDataPipe.functions:
+            if attribute_name in _deprecated_functional_names:
+                kwargs = _deprecated_functional_names[attribute_name]
+                _deprecation_warning(**kwargs)
```
Could you please add the similar logic to `MapDataPipe`?

ejguan reviewed (approved):
LGTM with two comments regarding `MapDataPipe`

NivekT committed:
Update on "[DataPipe] Add function for deprecation of functional DataPipe names"


Fixes https://github.com/pytorch/data/issues/480.

For the list of deprecated DataPipes (or functional names), see https://github.com/pytorch/data/issues/163

Testing:
```python
IterableWrapper(range(10)).open_file_by_iopath()
```
Returns:
```
/Users/.../pytorch/torch/utils/data/datapipes/utils/common.py:171: FutureWarning: `IoPathFileOpener()`'s functional API `.open_file_by_iopath()` is deprecated since 1.12 and will be removed in 1.14.
See https://github.com/pytorch/data/issues/163 for details.
Please use `.open_files_by_iopath()` instead.
  warnings.warn(msg, FutureWarning)
  ```

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/datapipe.py:
```diff
@@ -232,6 +237,10 @@ class MapDataPipe(Dataset[T_co], metaclass=_DataPipeMeta):
 
     def __getattr__(self, attribute_name):
         if attribute_name in MapDataPipe.functions:
+            if attribute_name in _deprecated_functional_names:
+                type, kwargs = _deprecated_functional_names[attribute_name]
```
Micro performance improvement. Why not split this `_deprecated_functional_names` to `_iter_deprecated_functional_names` and `_map_deprecated_functional_names` to reduce time to search the `attribute_name` in a single dictionary?

NivekT committed:
Update on "[DataPipe] Add function for deprecation of functional DataPipe names"


Fixes https://github.com/pytorch/data/issues/480.

For the list of deprecated DataPipes (or functional names), see https://github.com/pytorch/data/issues/163

Testing:
```python
IterableWrapper(range(10)).open_file_by_iopath()
```
Returns:
```
/Users/.../pytorch/torch/utils/data/datapipes/utils/common.py:171: FutureWarning: `IoPathFileOpener()`'s functional API `.open_file_by_iopath()` is deprecated since 1.12 and will be removed in 1.14.
See https://github.com/pytorch/data/issues/163 for details.
Please use `.open_files_by_iopath()` instead.
  warnings.warn(msg, FutureWarning)
  ```

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Add function for deprecation of functional DataPipe names"


Fixes https://github.com/pytorch/data/issues/480.

For the list of deprecated DataPipes (or functional names), see https://github.com/pytorch/data/issues/163

Testing:
```python
IterableWrapper(range(10)).open_file_by_iopath()
```
Returns:
```
/Users/.../pytorch/torch/utils/data/datapipes/utils/common.py:171: FutureWarning: `IoPathFileOpener()`'s functional API `.open_file_by_iopath()` is deprecated since 1.12 and will be removed in 1.14.
See https://github.com/pytorch/data/issues/163 for details.
Please use `.open_files_by_iopath()` instead.
  warnings.warn(msg, FutureWarning)
  ```

[ghstack-poisoned]

NivekT commented:
@pytorchbot merge -g

NivekT commented on torch/utils/data/datapipes/utils/common.py:
```diff
@@ -104,6 +104,25 @@ def validate_pathname_binary_tuple(data: Tuple[str, IOBase]):
+                                                       "deprecate_functional_name_only": True},
+                                                      "open_file_by_iopath":
+                                                      {"old_class_name": "IoPathFileOpener",
+                                                       "deprecation_version": "1.12",
+                                                       "removal_version": "1.14",
```
This has been fixed https://github.com/pytorch/pytorch/pull/79302
